GLTFLoader: Implement support for EXT_meshopt_compression

zeux opened the pull request:
EXT_meshopt_compression is a recent extension that supports compressing
glTF data - geometry, animation, instance transforms - in a generic fashion.

Specification:
https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_meshopt_compression

It is possible to produce files that use the extension using gltfpack
(https://github.com/zeux/meshoptimizer/tree/master/gltf). To decompress
the files, loader needs to decode the bufferView data - the extension
specification details the compressed format, but meshoptimizer library
provides a portable decoder that can be readily consumed from JS; the
decoder requires WebAssembly and will automatically dispatch to Wasm SIMD
when available, or use MVP otherwise.

meshopt_decoder.js is a drop-in copy of
https://github.com/zeux/meshoptimizer/blob/master/js/meshopt_decoder.js;
since it implements the specified format, it can be updated at any point
without loss of backwards compatibility.

The PR for EXT_meshopt_compression in glTF repository
(https://github.com/KhronosGroup/glTF/pull/1830) contains file sizes for
a lot of example models - the benefit of using this extension is that it
supports compressing every possible type of glTF binary data except for
images, and can decode at breakneck speeds (~1 GB/s when using Wasm SIMD
on a recent desktop Intel CPU) while only requiring a very small decoder
binary (meshopt_decoder.js is at 21 KB raw, 6 KB deflated, while
including *two* builds of the decoder, scalar and SIMD).

As an example, I've tested this PR on BrainStem model from
glTF-Sample-Models repository that contains animation data. The original
file size is 3120 KB, Draco-compressed version is 1942 KB, and
meshopt-compressed version is 362 KB, produced by gltfpack with `-cc`
option. The large size delta is due to animation data being compressed
with the extension; the loader extension decompresses bufferViews upon
request and the rest of the loader pipeline works without further
changes.

To integrate the extension after this change the user simply needs to
include the module, and pass the exported MeshoptDecoder global to the
loader:

	import './js/libs/meshopt_decoder.js';
	...
	loader.setMeshoptDecoder( MeshoptDecoder );

zeux committed:
GLTFLoader: Implement support for EXT_meshopt_compression

EXT_meshopt_compression is a recent extension that supports compressing
glTF data - geometry, animation, instance transforms - in a generic
fashion.

Specification:
https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_meshopt_compression

It is possible to produce files that use the extension using gltfpack
(https://github.com/zeux/meshoptimizer/tree/master/gltf). To decompress
the files, loader needs to decode the bufferView data - the extension
specification details the compressed format, but meshoptimizer library
provides a portable decoder that can be readily consumed from JS; the
decoder requires WebAssembly and will automatically dispatch to Wasm SIMD
when available, or use MVP otherwise.

meshopt_decoder.js is a drop-in copy of
https://github.com/zeux/meshoptimizer/blob/master/js/meshopt_decoder.js;
since it implements the specified format, it can be updated at any point
without loss of backwards compatibility.

The PR for EXT_meshopt_compression in glTF repository
(https://github.com/KhronosGroup/glTF/pull/1830) contains file sizes for
a lot of example models - the benefit of using this extension is that it
supports compressing every possible type of glTF binary data except for
images, and can decode at breakneck speeds (~1 GB/s when using Wasm SIMD
on a recent desktop Intel CPU) while only requiring a very small decoder
binary (meshopt_decoder.js is at 21 KB raw, 6 KB deflated, while
including *two* builds of the decoder, scalar and SIMD).

As an example, I've tested this PR on BrainStem model from
glTF-Sample-Models repository that contains animation data. The original
file size is 3120 KB, Draco-compressed version is 1942 KB, and
meshopt-compressed version is 362 KB, produced by gltfpack with `-cc`
option. The large size delta is due to animation data being compressed
with the extension; the loader extension decompresses bufferViews upon
request and the rest of the loader pipeline works without further
changes.

To integrate the extension after this change the user simply needs to
include the module, and pass the exported MeshoptDecoder global to the
loader:

	import './js/libs/meshopt_decoder.js';

	...

	loader.setMeshoptDecoder( MeshoptDecoder );

zeux committed:
Revert accidental whitespace change

zeux commented:
Note: based on the requests for my previous extension PRs I've omitted the example code/data; if you'd like to test this yourself and don't want to use gltfpack, I'm attaching a .zip with a .glb file (that should go to models/gltf) and an .html file (that's a quick edit of the existing animation example) for reference.

If you do want to use gltfpack to play around with this, note that this requires a recent gltfpack build, you can grab one from the artifacts here https://github.com/zeux/meshoptimizer/actions/runs/302546580 - gltfpack 0.15 will be released later this month on NPM that adheres to the final extension spec.

[meshopt-test.zip](https://github.com/mrdoob/three.js/files/5375146/meshopt-test.zip)

zeux committed:
When the decoder isn't specified or WebAssembly isn't supported, skip decoding

This makes sure that files with EXT_meshopt_compression used as a
non-required extension don't fail to load.

zeux committed:
Emit proper error messages when extension is required

Long term we might want to have a way to indicate failure from a plugin,
but for now this seems reasonable.

zeux commented:
The lint error is
> Error: src/cameras/CubeCamera.d.ts(16,2): error TS2416: Property 'clear' in type 'CubeCamera' is not assignable to the same property in base type 'Object3D'.

Which doesn't look related to this change, so presumably a different change was merged that introduced this.

donmccurdy commented:
Thanks @zeux! Excited to have this option available by default.

> import './js/libs/meshopt_decoder.js';

@gkjohnson does this import method ([side effects only](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Import_a_module_for_its_side_effects_only)) raise any red flags for issues you know of? I guess I'm wondering, if we're distributing the file with three.js so that it'd be imported from `three/examples/js/lib/*`, and we later want to put `"sideEffects": false` in our `package.json`, is it going to break?

For comparison, we made ZSTDDecoder a module (https://github.com/mrdoob/three.js/pull/19932) and didn't create a global/UMD version.

> Note: based on the requests for my previous extension PRs I've omitted the example code/data;

At some point I think `EXT_meshopt_compression` should be included in one of the "extensions" examples: https://threejs.org/examples/?q=gltf#webgl_loader_gltf_extensions. Perhaps just adding a single additional version of one of the models already there, for QA testing purposes — it doesn't have to be included with this PR, though.

> Which doesn't look related to this change, so presumably a different change was merged that introduced this.

Agreed, the lint error can be ignored.

zeux commented:
FWIW the reason why the import is state-modifying is that the .js file in question has a UMD-style declaration, so you can load it in a node environment using require, in a browser environment pre-ES imports using any of the existing import libraries, in a browser environment *without* any import libraries by just using `<script>` tag, or using an ES-style import. Unfortunately you can not export a symbol conditionally which forced me to resort to the solution that's present here.

The alternative here could be to change the UMD-style export at the end of meshopt_decoder.js to a ES6 export statement, which would force the user to use modules.

zeux commented:
I can move the decoder js to `jsm/libs/meshopt_decoder.module.js` and change UMD to an export; this would require ES6 imports to use. The original plan for modules, I think, was to drop the plain JS files, but the issue linked suggests that it may not be the plan anymore, so I'm slightly concerned about this - is that something to worry about?

Alternatively I can *add* a module version to jsm/libs and keep the original one?

gkjohnson commented:
> The original plan for modules, I think, was to drop the plain JS files, but the issue linked suggests that it may not be the plan anymore, so I'm slightly concerned about this

I'm not exactly sure what the plan is -- it was left a bit ambiguous beyond the plan being to transition to build the `/jsm` directories to `/js` rather than the reverse which happens now. Of course I don't have the final say on how all this should work but I feel that at least a `/jsm` variant should be provided. It would be consistent with the way `ZSTDDecoder` works, as well.

zeux committed:
Add meshopt_decoder.module.js

This allows to use a ES6 style import that doesn't modify globals, e.g.

import { MeshoptDecoder } from './jsm/libs/meshopt_decoder.module.js';

zeux commented:
Updated the PR with a module version of the decoder that uses ES6 export, so you can import like this:

```
import { MeshoptDecoder } from './jsm/libs/meshopt_decoder.module.js';
```

donmccurdy commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -261,6 +276,23 @@ THREE.GLTFLoader = ( function () {
+
+								}
+
+							}
```
This doesn't seem like the right place for these checks (this section contains extensions we haven't yet migrated to the new extension API) but I don't immediately know what the right place would be. Does anything break other than less clear user warnings if they're removed? Or we could keep them here and figure out what to do later.

zeux replied:
This is purely for error clarity - with these removed, you get this:

RangeError: Invalid typed array length: 1048
    at new Float32Array (<anonymous>)
    at GLTFLoader.js:2291

I'm okay with omitting this for now? I also just now realized that BasisU integration suffers from an opposite problem - I've added a similar check to the extension code in my original PR, but this means that if the loader isn't configured, a glTF file with an *optional* extension (and a fallback non-compressed texture) will fail to load if you don't register the KTX loader.

I could omit these error messages from this PR and submit a fix to KTX loader (that will result in poor error handling behavior but it's better than incorrectly failing to load files), and then we could revisit this somehow?

donmccurdy replied:
I guess this is just something we need to address in the extension API, it seems OK to keep this PR as it is for now. Thanks!

donmccurdy commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -261,6 +276,23 @@ THREE.GLTFLoader = ( function () {
+						case EXTENSIONS.EXT_MESHOPT_COMPRESSION:
+							if ( extensionsRequired.includes( extensionName ) ) {
+
+								if ( !this.meshoptDecoder ) {
```
nit: space between `! this.meshoptDecoder` here and below, I think the linter will complain otherwise.

donmccurdy reviewed (commented):
> I'm not exactly sure what the plan is -- it was left a bit ambiguous beyond the plan being to transition to build the /jsm directories to /js rather than the reverse which happens now.

I'm not planning to add any new files to `/js`, but to add new features to `/jsm` instead. So KTX2Loader and ZSTDDecoder would remain available as ES modules, only. If we do set up automation that generates `/jsm` -> `/js` then those files could perhaps be included in that, TBD.

zeux committed:
Fix code style and missing break

zeux committed:
Merge branch 'dev' into meshopt-compression

zeux committed:
Replace error handling with an extension-local solution

Instead of changing the extension parsing loop, we now use the fallback
only if extension is not present in the required list.

zeux commented:
@mrdoob @donmccurdy lmk if I need to do anything else here

donmccurdy commented:
No concerns here, functionally looks correct to me. 👍 

@mrdoob if you'd prefer to stick with the earlier decision of not adding new libraries in `examples/js`, I think we could also release this with just the `examples/jsm` dependency for now and (optionally) generate the `examples/js` version later on.

mrdoob commented:
> @mrdoob if you'd prefer to stick with the earlier decision of not adding new libraries in `examples/js`, I think we could also release this with just the `examples/jsm` dependency for now and (optionally) generate the `examples/js` version later on.

Nah, I think adding new libraries in `examples/js` in this case is good.

mrdoob commented:
Ops. Seems like the `EXT_texture_webp` PR created conflicts... @zeux could you resolve them?

zeux committed:
Merge https://github.com/mrdoob/three.js into meshopt-compression

mrdoob commented:
Thanks!
